### DISCUSSION TITLE
NVM pre-installed on macOS only?

### DISCUSSION THREAD

**Discussion by scottohara** [community/community#25885] [Actions]
<p>I’ve started migrating a workflow from TravisCI to GitHub actions.</p>
<p>The workflow is for a node.js app, and the repo includes an <code>.nvmrc</code> file that specifies the version of Node to install/use.</p>
<p>On TravisCI (where <code>nvm</code> is pre-installed), I previously used the following command to install the desired node.js version:</p>
<pre><code>before_install:
  - nvm install
</code></pre>
<p>Unfortunately I can’t do a similar thing in GitHub actions, because (as per the <a href="https://help.github.com/en/articles/software-in-virtual-environments-for-github-actions" rel="nofollow noopener">software in virtual environments</a> page) it seems that <code>nvm</code> only comes pre-installed for the macOS environment, but not the Ubuntu environments.  Was there a specific reason for not having NVM available on Ubuntu?</p>
<p>From what I understand, in Github Actions the common practice for installing a different Node.js version (than the default pre-installed one) is to use the <code>setup-node</code> action and specify the version to install:</p>
<pre><code>steps:
-uses:actions/setup-node@v1with:
node-version:'10.x'
</code></pre>
<p>Unfortunately, <code>setup-node</code> currently does not look for a <code>.nvmrc</code> file (but it looks like <a href="https://github.com/actions/setup-node/issues/32#issuecomment-522083355" rel="nofollow noopener">support for this is being considered</a>), and the workaround (read the contents of <code>.nvmrc</code> into an output) seems like a bit of an ugly hack:</p>
<pre><code>- name: Read .nvmrc
  run: echo "##[set-output name=NVMRC;]$(cat .nvmrc)"
  id: nvm

- name: Use Node.js (.nvmrc)
  uses: actions/setup-node@v1
  with:
    node-version: "${{ steps.nvm.outputs.NVMRC }}"
</code></pre>
<p>I’m interested to understand if there was a reason why NVM was picked as a pre-installed option on macOS only, and not on any other environments.</p>

**Comment by scottohara** [comment]
<p>For what it’s worth, I tried adding an “Install nvm” step into my workflow as follows:</p>
<pre><code>- name: Install nvm
  run: |
    curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
    export NVM_DIR="$HOME/.nvm"
    source "$NVM_DIR/nvm.sh"
</code></pre>
<p>The <code>curl</code> command runs fine (I tested it on its own first); but the three commands together fail with:</p>
<pre><code>##[error]Process completed with exit code 3.
</code></pre>
<p>The full output of the step is below:</p>
<pre><code>curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
  export NVM_DIR="$HOME/.nvm"
  source "$NVM_DIR/nvm.sh"
  shell: /bin/bash -e {0}
  % Total % Received % Xferd Average Speed Time Time Time Current
                                 Dload Upload Total Spent Left Speed

  0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
100 13226 100 13226 0 0 92489 0 --:--:-- --:--:-- --:--:-- 92489
=&gt; Downloading nvm from git to '/home/runner/.nvm'

Cloning into '/home/runner/.nvm'...
=&gt; =&gt; Compressing and cleaning up git repository

=&gt; Appending nvm source string to /home/runner/.bashrc
=&gt; Appending bash_completion source string to /home/runner/.bashrc
N/A: version "12.1.0 -&gt; N/A" is not yet installed.

You need to run "nvm install 12.1.0" to install it before using it.
=&gt; You currently have modules installed globally with `npm`. These will no
=&gt; longer be linked to the active version of Node when you install a new node
=&gt; with `nvm`; and they may (depending on how you construct your `$PATH`)
=&gt; override the binaries of modules installed with `nvm`:

/usr/local/lib
├── bower@1.8.8
├── grunt@1.0.4
├── gulp@4.0.2
├── n@6.0.1
├── now@16.1.1
├── parcel-bundler@1.12.3
├── typescript@3.5.3
├── webpack@4.39.2
└── webpack-cli@3.3.7
=&gt; If you wish to uninstall them at a later point (or re-install them under your
=&gt; `nvm` Nodes), you can remove them from the system Node as follows:

     $ nvm use system
     $ npm uninstall -g a_module

=&gt; Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[-s "$NVM_DIR/nvm.sh"] &amp;&amp; \. "$NVM_DIR/nvm.sh" # This loads nvm
[-s "$NVM_DIR/bash_completion"] &amp;&amp; \. "$NVM_DIR/bash_completion" # This loads nvm bash_completion
##[error]Process completed with exit code 3.
</code></pre>
<p>Not entirely sure how to diagnose the “process completed with exit code 3” error?</p>

**Comment by ethomson** [comment, marked as answer]
<p>nvm is <a href="https://github.com/nvm-sh/nvm/blob/master/nvm.sh#L2944" rel="nofollow noopener">returning error code 3</a>.  There’s several reports about this in their issue tracker, in particular, one that sounds relevant describing this behavior when <a href="https://github.com/nvm-sh/nvm/issues/1985#issuecomment-456022013" rel="nofollow noopener">installing nvm while currently in a directory that contains a .nvmrc file</a>.</p>
<p>There are some potential workarounds there like passing the <code>--install</code> flag.</p>

**Comment by scottohara** [comment]
<p>Thanks <a class="mention" href="/ethomson">@ethomson</a> , I followed your suggestion and was able to get it working with the <code>--install</code> flag.</p>
